src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,189 @@
+        {
+            _writeEvent(evt);
+        }
+
+        private void WriteToConsoleBuffer(string evt)
+        {
+            try
+            {
+                while (_consoleBuffer.Writer.TryWrite(evt) == false)
+                {
+                    // Buffer is currently full, wait until writing is permitted.
+                    // This is the downside of using channels, we are on a sync code path and so we have to block on this task
+                    var writeTask = _consoleBuffer.Writer.WaitToWriteAsync().AsTask();
+                    if (writeTask.WaitAsync(_consoleBufferTimeout).Result == false)
+                    {
+                        // The buffer has been completed and does not allow further writes.
+                        // Currently this should not be possible, but the safest thing to do is just write directly to the console.
+                        Console.WriteLine(evt);
+                        break;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                // Most likely a task cancellation exception from the timeout expiring, but regardless we use the handler
+                // dump the raw exception and write the event to the console directly
+                _exceptionhandler(ex);
+                Console.WriteLine(evt);
+            }
+        }
+
+        internal void StartProcessingBuffer(bool batched)
+        {
+            // intentional no-op if the task is already running
+            if (_consoleBufferReadLoop == null || _consoleBufferReadLoop.IsCompleted)
+            {
+                _consoleBufferReadLoop = ProcessConsoleBufferAsync(batched);
+            }
+        }
+
+        private async Task ProcessConsoleBufferAsync(bool batched)
+        {
+            try
+            {
+                if (batched)
+                {
+                    await ProcessConsoleBufferBatchedAsync();
+                }
+                else
+                {
+                    await ProcessConsoleBufferNonBatchedAsync();
+                }
+            }
+            catch (Exception ex)
+            {
+                // Not sure what can fail here, but there is nothing else monitoring this task so just log the exception
+                _exceptionhandler(ex);
+            }
+            finally
+            {
+                // if this has failed for any reason, fall everything back to console
+                _writeEvent = Console.WriteLine;
+                _consoleBuffer.Writer.Complete();
+            }
+        }
+
+        private async Task ProcessConsoleBufferNonBatchedAsync()
+        {
+            await foreach (var line in _consoleBuffer.Reader.ReadAllAsync())
+            {
+                Console.WriteLine(line);
+            }
+        }
+
+        private async Task ProcessConsoleBufferBatchedAsync()
+        {
+            var builder = new StringBuilder();
+
+            while (true)
```

Comment:
nit: WaitToReadyAsync can be the directly in the outer `while` statement, and then an inner loop can be used to rapidly deque items without performing another wait:

``` CSharp
StringBuilder builder = new();
while (await reader.WaitToReadAsync())
{
    int lines = 0;
    string line = null;
    while (lines < DefaultBufferSize && reader.TryRead(out string item))
    {
        lines++;
        switch (lines)
        {
            case 0:
                line = item;
                break;
            case 1:
                builder.AppendLine(line);
                builder.AppendLine(item);
                break;
            default:
                builder.AppendLine(item);
                break;
        }
    }

    if (lines == 1)
    {
        Console.WriteLine(line);
    }
    else if (lines > 1)
    {
       Console.WriteLine(builder.ToString());
       builder.Clear();
    }
}
```